packages/turbo/bin/turbo

```diff
@@ -1,10 +1,257 @@
+  const otherInstalled = definitelyWrongBinaries.filter(binaryPath => {
+    try {
+      return require.resolve(binaryPath);
+    } catch (e) {}
+  });
+
+  console.error();
+  console.error('***');
+  console.error();
+
+  if (otherInstalled.length > 0) {
+    console.error('Turborepo checked to see if binaries for another platform are installed.');
+    console.error('This typically indicates an error in sharing of pre-resolved node_modules across platforms.');
+    console.error('One common reason for this is copying files to Docker.');
+    console.error();
+    console.error(`We found these unnecessary binaries:`);
+    console.error(otherInstalled.join('\n'));
+  } else {
+    console.error(`We did not find any binaries on this system.`);
+    console.error(`This can happen if you run installation with the --no-optional flag.`);
+  }
+
+  // Check to see if we have partially-populated dependencies in the npm lockfile.
+  const MAX_LOOKUPS = 10;
+  const availablePackages = availablePlatforms.map(platform => availableArchs.map(arch => `turbo-${platform}-${arch}`)).flat();
+
+  try {
+    // Attempt to find project root.
+    const selfPath = require.resolve('turbo/package');
+
+    let previous = null;
+    let current = path.join(selfPath, '..', 'package-lock.json');
+
+    for (let i = 0; previous !== current && i < MAX_LOOKUPS; i++) {
+      try {
+        const lockfile = fs.readFileSync(current);
+        const parsedLockfile = JSON.parse(lockfile);
+
+        // If we don't show up in the lockfile it's the wrong lockfile.
+        if (parsedLockfile?.dependencies?.turbo) {
+          // Check to see if all of `turbo-<PLATFORM>-<ARCH>` is included.
+          const hasAllPackages = Object.keys(parsedLockfile?.dependencies ?? {}).filter(dependency => availablePackages.includes(dependency)).length === availablePackages.length;
+          if (!hasAllPackages) {
+            console.error();
+            console.error('***');
+            console.error();
+            console.error(`Turborepo detected that your lockfile (${current}) does not enumerate all available platforms.`);
+            console.error('This is likely a consequence of an npm issue: https://github.com/npm/cli/issues/4828.');
+
+            // Let's build their repair command:
+            let version = '';
+            let environment = '';
+            if (parsedLockfile?.packages[""]?.dependencies?.turbo) {
+              version = `@${parsedLockfile.packages[""].dependencies.turbo}`;
+              environment = ' --save-prod';
+            } else if (parsedLockfile?.packages[""]?.devDependencies?.turbo) {
+              version = `@${parsedLockfile.packages[""].devDependencies.turbo}`;
+              environment = ' --save-dev';
+            } else if (parsedLockfile?.packages[""]?.optionalDependencies?.turbo) {
+              version = `@${parsedLockfile.packages[""].optionalDependencies.turbo}`;
+              environment = ' --save-optional';
+            }
```

Comment:
@chris-olszewski same note here:

I'm trying to find the version of `turbo` in use, with the range qualifier. Are these JSON paths going to do what I want them to do?

Reply:
Yup these will work for npm 7+ lockfiles. Lockfile version 1 won't have `packages` and we'd need to check `(dev|opt)?(d|D)ependencies` in the root `package.json` itself.

Reply:
Okay, I'm going to elect against perfect support for `npm@6`. They'll get a less-perfect repair command that will still work but may cause more changes in their lockfile than strictly necessary.